lib/phoenix_live_view/tag_engine.ex

```diff
@@ -1160,6 +1163,93 @@ defmodule Phoenix.LiveView.TagEngine do
     end
   end
 
+  defp handle_keyed_comprehension(key_expr, state, tag_meta) do
+    # for now, we only support plain tags because we don't know if a function component
+    # renders to a single tag which is required by live components
+    case state.tag_handler.classify_type(tag_meta.tag_name) do
+      {:tag, _} ->
+        :ok
+
+      other ->
+        type =
+          case other do
+            {:local_component, _} -> "function components"
+            {:remote_component, _} -> "function components"
+            # special attrs for slots are handled separately, see wrap_special_slot
+            other -> inspect(other)
+          end
+
+        raise_syntax_error!(
+          "keyed :for comprehensions are not supported on #{type}",
+          tag_meta,
+          state
+        )
+    end
+
+    for_expr =
+      tag_meta[:for] || raise_syntax_error!("cannot use :key without :for", tag_meta, state)
+
+    if_expr = tag_meta[:if]
+    for_assigns = extract_variables(for_expr)
+    vars_changed = Enum.map(for_assigns, fn {name, _} -> name end)
+
+    ast =
+      invoke_subengine(state, :handle_body, [
+        [skip_require: true, root: true, vars_changed: vars_changed]
+      ])
+
+    vars_changed = Macro.var(:vars_changed, Phoenix.LiveView.Engine)
+
+    new_ast =
+      quote do
+        Phoenix.LiveView.TagEngine.component(
+          &Phoenix.Component.live_component/1,
+          %{
+            id: {unquote(state.caller.module), unquote(tag_meta.line), unquote(key_expr)},
+            module: Phoenix.LiveView.KeyedComprehension,
+            vars_changed: %{unquote_splicing(for_assigns)},
+            render: fn unquote(vars_changed) -> unquote(ast) end
+          },
+          {__MODULE__, __ENV__.function, __ENV__.file, unquote(tag_meta.line)}
+        )
+      end
+
+    state = pop_substate_from_stack(state)
+
+    for_ast =
+      state
+      |> push_substate_to_stack()
+      |> update_subengine(:handle_begin, [])
+      |> update_subengine(:handle_expr, ["=", new_ast])
+      |> invoke_subengine(:handle_end, [])
+
+    if if_expr do
+      quote do
+        for unquote(for_expr), unquote(if_expr), do: unquote(for_ast)
+      end
+    else
+      quote do
+        for unquote(for_expr), do: unquote(for_ast)
+      end
+    end
+  end
+
+  # we already validated that the for expression has the correct shape in
+  # validate_quoted_special_attr
+  defp extract_variables({:<-, _for_meta, [lhs, _rhs]}) do
+    {_ast, variables} =
+      Macro.prewalk(lhs, [], fn
```

Comment:
There are some variables we should ignore, such as `^foo` or on the right side of `::`. It should be a matter of adding these clauses:

```
{:^, _, [expr]}, acc ->
  {[], acc}

{:"::", _, [left, right]}, acc ->
  {left, acc}
```